### PR TITLE
[Darwin] MTRDevice attribute storage mutable array enumeration fix

### PR DESCRIPTION
**Pull request by jtung-apple**
This fixes the bug where the mutable array is modified during enumeration.

### CODE REVIEW

File: src/darwin/Framework/CHIP/MTRDeviceControllerDataStore.mm [hunk 1]
```diff
@@ -499,79 +525,95 @@ - (void)_pruneEmptyStoredAttributesBranches
...
-        NSMutableArray<NSNumber *> * endpointIndex = [self _fetchEndpointIndexForNodeID:nodeID].mutableCopy;
-        NSUInteger endpointIndexCount = endpointIndex.count;
+        NSArray<NSNumber *> * endpointIndex = [self _fetchEndpointIndexForNodeID:nodeID];
+        NSMutableArray<NSNumber *> * endpointIndexCopy = endpointIndex.mutableCopy;
 
         for (NSNumber * endpointID in endpointIndex) {
             // Fetch endpoint index
-            NSMutableArray<NSNumber *> * clusterIndex = [self _fetchClusterIndexForNodeID:nodeID endpointID:endpointID].mutableCopy;
-            NSUInteger clusterIndexCount = clusterIndex.count;
+            NSArray<NSNumber *> * clusterIndex = [self _fetchClusterIndexForNodeID:nodeID endpointID:endpointID].mutableCopy;
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/MTRDeviceControllerDataStore.mm:541]:
```suggestion
            NSArray<NSNumber *> * clusterIndex = [self _fetchClusterIndexForNodeID:nodeID endpointID:endpointID];
```

File: src/darwin/Framework/CHIP/MTRDeviceControllerDataStore.mm [hunk 2]
```diff
@@ -499,79 +525,95 @@ - (void)_pruneEmptyStoredAttributesBranches
...
+            NSMutableArray<NSNumber *> * clusterIndexCopy = clusterIndex.mutableCopy;
 
             for (NSNumber * clusterID in clusterIndex) {
                 // Fetch endpoint index
-                NSMutableArray<NSNumber *> * attributeIndex = [self _fetchAttributeIndexForNodeID:nodeID endpointID:endpointID clusterID:clusterID].mutableCopy;
-                NSUInteger attributeIndexCount = attributeIndex.count;
+                NSArray<NSNumber *> * attributeIndex = [self _fetchAttributeIndexForNodeID:nodeID endpointID:endpointID clusterID:clusterID].mutableCopy;
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/MTRDeviceControllerDataStore.mm:546]:
```suggestion
                NSArray<NSNumber *> * attributeIndex = [self _fetchAttributeIndexForNodeID:nodeID endpointID:endpointID clusterID:clusterID];
```

File: src/darwin/Framework/CHIP/MTRDeviceControllerDataStore.mm [hunk 3]
```diff
@@ -499,79 +525,95 @@ - (void)_pruneEmptyStoredAttributesBranches
...
                     }
                 }
 
-                if (!attributeIndex.count) {
-                    [clusterIndex removeObject:clusterID];
-                } else if (attributeIndex.count != attributeIndexCount) {
-                    BOOL success = [self _storeAttributeIndex:attributeIndex forNodeID:nodeID endpointID:endpointID clusterID:clusterID];
+                if (attributeIndex.count != attributeIndexCopy.count) {
+                    BOOL success;
                     if (!success) {
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/MTRDeviceControllerDataStore.mm]: This doesn't make sense.  `success` is not even initialized here!

File: src/darwin/Framework/CHIP/MTRDeviceControllerDataStore.mm [hunk 4]
```diff
@@ -499,79 +525,95 @@ - (void)_pruneEmptyStoredAttributesBranches
...
+                        if (attributeIndexCopy.count) {
+                            success = [self _storeAttributeIndex:attributeIndexCopy forNodeID:nodeID endpointID:endpointID clusterID:clusterID];
+                        } else {
+                            [clusterIndexCopy removeObject:clusterID];
+                            success = [self _deleteAttributeIndexForNodeID:nodeID endpointID:endpointID clusterID:clusterID];
+                        }
                         storeFailures++;
-                        MTR_LOG_INFO("Store failed for attributeIndex");
+                        MTR_LOG_INFO("Store failed in _pruneEmptyStoredAttributesBranches for attributeIndex @ 0x%016llX:0x%04X:0x%08lX", nodeID.unsignedLongLongValue, endpointID.unsignedShortValue, clusterID.unsignedLongValue);
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/MTRDeviceControllerDataStore.mm]: Again, decimal endpoints.  Will stop commenting on that; just fix all of them, please.

File: src/darwin/Framework/CHIP/MTRDeviceControllerDataStore.mm [hunk 5]
```diff
@@ -499,79 +525,95 @@ - (void)_pruneEmptyStoredAttributesBranches
...
                 }
             }
 
-            if (!clusterIndex.count) {
-                [endpointIndex removeObject:endpointID];
-            } else if (clusterIndex.count != clusterIndexCount) {
-                BOOL success = [self _storeClusterIndex:clusterIndex forNodeID:nodeID endpointID:endpointID];
+            if (clusterIndex.count != clusterIndexCopy.count) {
+                BOOL success;
+                if (clusterIndex.count) {
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/MTRDeviceControllerDataStore.mm]: This is checking the wrong count.

File: src/darwin/Framework/CHIP/MTRDeviceControllerDataStore.mm [hunk 6]
```diff
@@ -499,79 +525,95 @@ - (void)_pruneEmptyStoredAttributesBranches
...
         }
     }
 
-    if (!nodeIndex.count) {
-        [self _deleteNodeIndex];
-    } else if (nodeIndex.count != nodeIndexCount) {
-        BOOL success = [self _storeNodeIndex:nodeIndex];
+    if (nodeIndex.count != nodeIndexCopy.count) {
+        BOOL success;
+        if (!nodeIndex.count) {
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/MTRDeviceControllerDataStore.mm]: This is checking the wrong count.

File: src/darwin/Framework/CHIP/MTRDeviceControllerDataStore.mm [hunk 7]
```diff
@@ -499,79 +525,95 @@ - (void)_pruneEmptyStoredAttributesBranches
...
         if (!success) {
             storeFailures++;
-            MTR_LOG_INFO("Store failed for nodeIndex");
+            MTR_LOG_INFO("Store failed in _pruneEmptyStoredAttributesBranches for nodeIndex");
         }
     }
 
     if (storeFailures) {
-        MTR_LOG_ERROR("Store failed in _pruneEmptyStoredAttributesBranches: %lu", (unsigned long) storeFailures);
+        MTR_LOG_ERROR("Store failed in _pruneEmptyStoredAttributesBranches: total %lu", static_cast<unsigned long>(storeFailures));
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/MTRDeviceControllerDataStore.mm:614]:
```suggestion
        MTR_LOG_ERROR("Store failed in _pruneEmptyStoredAttributesBranches: failure count %lu", static_cast<unsigned long>(storeFailures));
```

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]

File: src/darwin/Framework/CHIP/MTRDeviceControllerDataStore.mm [hunk 8]
```diff
@@ -495,83 +521,106 @@ - (void)_deleteAttributeValueForNodeID:(NSNumber *)nodeID endpointID:(NSNumber *
...
-            NSUInteger clusterIndexCount = clusterIndex.count;
+            NSArray<NSNumber *> * clusterIndex = [self _fetchClusterIndexForNodeID:nodeID endpointID:endpointID];
+            NSMutableArray<NSNumber *> * clusterIndexCopy = clusterIndex.mutableCopy;
 
             for (NSNumber * clusterID in clusterIndex) {
                 // Fetch endpoint index
-                NSMutableArray<NSNumber *> * attributeIndex = [self _fetchAttributeIndexForNodeID:nodeID endpointID:endpointID clusterID:clusterID].mutableCopy;
-                NSUInteger attributeIndexCount = attributeIndex.count;
+                NSArray<NSNumber *> * attributeIndex = [self _fetchAttributeIndexForNodeID:nodeID endpointID:endpointID clusterID:clusterID];
+                NSMutableArray<NSNumber *> * attributeIndexCopy = attributeIndex.mutableCopy;
```
Comment (ksperling-apple) [inline src/darwin/Framework/CHIP/MTRDeviceControllerDataStore.mm:556]:
```suggestion
                NSMutableArray<NSNumber *> * attributeIndexCopy = [attributeIndex mutableCopy];
```
`mutableCopy` is not semantically a property (the reason for calling it is to get a new, mutable copy), so it should really not be called using property syntax. (Same for all the other instances)

File: src/darwin/Framework/CHIP/MTRDeviceControllerDataStore.mm [hunk 9]
```diff
@@ -437,45 +437,71 @@ - (BOOL)_storeAttributeValue:(NSDictionary *)value forNodeID:(NSNumber *)nodeID
...
+
         if (![nodeIndex containsObject:nodeID]) {
             // Sanity check and delete if nodeID exists in index
             NSArray<NSNumber *> * endpointIndex = [self _fetchEndpointIndexForNodeID:nodeID];
             if (endpointIndex) {
                 MTR_LOG_ERROR("Persistent attribute cache contains orphaned entry for nodeID %@ - deleting", nodeID);
                 [self clearStoredAttributesForNodeID:nodeID];
             }
+
+            MTR_LOG_INFO("Fetch got no value for endpointIndex @ %llu", nodeID.unsignedLongLongValue);
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/MTRDeviceControllerDataStore.mm:470]: It looks like you changed all the _node ids_ to be logged as decimal and left the _endpoint ids_ as hex?  That's exactly backwards.

File: src/darwin/Framework/CHIP/MTRDeviceControllerDataStore.mm [hunk 10]
```diff
@@ -495,83 +521,106 @@ - (void)_deleteAttributeValueForNodeID:(NSNumber *)nodeID endpointID:(NSNumber *
...
+                if (attributeIndex.count != attributeIndexCopy.count) {
+                    BOOL success;
+                    if (attributeIndexCopy.count) {
+                        success = [self _storeAttributeIndex:attributeIndexCopy forNodeID:nodeID endpointID:endpointID clusterID:clusterID];
+                    } else {
+                        [clusterIndexCopy removeObject:clusterID];
+                        success = [self _deleteAttributeIndexForNodeID:nodeID endpointID:endpointID clusterID:clusterID];
                     }
+                    storeFailures++;
+                    MTR_LOG_INFO("Store failed in _pruneEmptyStoredAttributesBranches for attributeIndex (%lu) @ %llu:0x%04X:0x%08lX", static_cast<unsigned long>(attributeIndexCopy.count), nodeID.unsignedLongLongValue, endpointID.unsignedShortValue, clusterID.unsignedLongValue);
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/MTRDeviceControllerDataStore.mm:575]: Shouldn't this be conditioned on !success?

File: src/darwin/Framework/CHIP/MTRDeviceControllerDataStore.mm (hunk 9 above)
Comment (jtung-apple) [reply src/darwin/Framework/CHIP/MTRDeviceControllerDataStore.mm:470]: I misread your original comment. I'll change it as originally intended.

**Review by bzbarsky-apple — APPROVED** [review]

**Review by yunhanw-google — APPROVED** [review]

**Pull request merged**